Src/AutoNSubstituteUnitTest/SubstituteAttributeRelayTest.cs

```diff
@@ -0,0 +1,163 @@
+            // Teardown
+        }
+
+        [Fact]
+        public void CreateGetsInheritedAttributesFromProviderBecauseAttributesAreInheritedByDefault()
+        {
+            // Fixture setup
+            var sut = new SubstituteAttributeRelay();
+            var request = Substitute.For<ICustomAttributeProvider>();
+            var context = Substitute.For<ISpecimenContext>();
+            // Exercise system
+            sut.Create(request, context);
+            // Verify outcome
+            request.Received().GetCustomAttributes(typeof(SubstituteAttribute), true);
+            // Teardown
+        }
+
+        [Fact]
+        public void CreateReturnsSpecimenResolvedByContextFromSubstituteRequest()
+        {
+            // Fixture setup
+            var sut = new SubstituteAttributeRelay();
+            var request = Substitute.For<ParameterInfo>();
+            request.ParameterType.Returns(typeof(IInterface));
+            request.GetCustomAttributes(Arg.Any<Type>(), Arg.Any<bool>()).Returns(new[] { new SubstituteAttribute() });
+            var expectedSpecimen = new object();
+            var context = Substitute.For<ISpecimenContext>();
+            context.Resolve(Arg.Any<SubstituteRequest>()).Returns(expectedSpecimen);
+            // Exercise system
+            object actualSpecimen = sut.Create(request, context);
+            // Verify outcome
+            Assert.Same(expectedSpecimen, actualSpecimen);
+            // Teardown
+        }
+
+        [Fact]
+        public void CreateResolvesSubstituteRequestForParameterWithSubstituteAttribute()
+        {
+            // Fixture setup
+            var sut = new SubstituteAttributeRelay();
+            var request = Substitute.For<ParameterInfo>();
+            request.ParameterType.Returns(typeof(IInterface));
+            request.GetCustomAttributes(Arg.Any<Type>(), Arg.Any<bool>()).Returns(new[] { new SubstituteAttribute() });
+            var context = Substitute.For<ISpecimenContext>();
+            // Exercise system
+            sut.Create(request, context);
+            // Verify outcome
+            context.Received().Resolve(Arg.Is<SubstituteRequest>(r => r.TargetType == request.ParameterType));
+            // Teardown
+        }
+
+        [Fact]
+        public void CreateResolvesSubstituteRequestForPropertyWithSubstituteAttribute()
+        {
+            // Fixture setup
+            var sut = new SubstituteAttributeRelay();
+            var request = Substitute.For<PropertyInfo>();
+            request.PropertyType.Returns(typeof(IInterface));
+            request.GetCustomAttributes(Arg.Any<Type>(), Arg.Any<bool>()).Returns(new[] { new SubstituteAttribute() });
+            var context = Substitute.For<ISpecimenContext>();
+            // Exercise system
+            sut.Create(request, context);
+            // Verify outcome
+            context.Received().Resolve(Arg.Is<SubstituteRequest>(r => r.TargetType == request.PropertyType));

```

Comment:
Given this test, I can still sabotage the implementation like this:

```C#
public object Create(object request, ISpecimenContext context)
{
    if (context == null)
    {
        throw new ArgumentNullException("context");
    }

    var customAttributeProvider = request as ICustomAttributeProvider;
    if (customAttributeProvider == null)
    {
        return new NoSpecimen(request);
    }

    var attribute = customAttributeProvider.GetCustomAttributes(typeof(SubstituteAttribute), true)
            .OfType<SubstituteAttribute>().FirstOrDefault();
    if (attribute == null)
    {
        return new NoSpecimen(request);
    }

    object substituteRequest = CreateSubstituteRequest(customAttributeProvider, attribute);
    if(customAttributeProvider is ParameterInfo)
        return context.Resolve(substituteRequest);
    context.Resolve(substituteRequest);
    return context.Resolve(new SubstituteRequest(typeof(object)));
}
```

Notice the last four lines of code, which are obviously malicious and incorrect.